xos.h

```diff
@@ -101,36 +102,44 @@ namespace xos
     int lookahead_;
   };
 
-} // namespace xos
+  class StrAST
+  {
+  public:
+    std::string StrVal;
 
-class StrAST
-{
-  std::string StrVal;
+    StrAST(const std::string &StrVal) : StrVal(StrVal) {}
+  };
 
-public:
-  StrAST(const std::string &StrVal) : StrVal(StrVal) {}
-};
+  class PrototypeAST
+  {
+  public:
+    std::string Name;
+    std::vector<std::string> Args;
+    std::string returnType;
+    PrototypeAST(const std::string &Name, std::vector<std::string> Args, std::string returnType)
+        : Name(Name), Args(std::move(Args)), returnType(std::move(returnType)) {}
 
-class PrototypeAST
-{
-  std::string Name;
-  std::vector<std::string> Args;
-  std::string returnType;
+    const std::string &getName() const { return Name; }
+  };
 
-public:
-  PrototypeAST(const std::string &Name, std::vector<std::string> Args, std::string returnType)
-      : Name(Name), Args(std::move(Args)), returnType(std::move(returnType)) {}
+  class FunctionAST
+  {
+  public:
+    std::unique_ptr<PrototypeAST> Proto;
+    std::unique_ptr<StrAST> Body;
+    FunctionAST(std::unique_ptr<PrototypeAST> Proto,
+                std::unique_ptr<StrAST> Body)
+        : Proto(std::move(Proto)), Body(std::move(Body)) {}
+  };
 
-  const std::string &getName() const { return Name; }
-};
+  class Parser
+  {
+  public:
+    Lexer lexer_; // [TODO] make this unique_ptr?
+    Parser(std::istream &input) : lexer_(input) {}
+    void parse();
```

Comment:
The highest level parse function should also return a node.

Also might want to instead have the various parse methods return `Result<std::unique_ptr<Node>>` since we may want to propagate error messages from the lexer.